### DISCUSSION TITLE
Minimal field permissions for uploading files?

### DISCUSSION THREAD

**Discussion by andy1li** [directus/directus#14979] [Archive]
I wonder whether the documentation on uploading files could be improved slightly? 

Specifically, it would be nice to have more clarity on the minimal field permissions that have to be set.

The goal is to grant as little field permissions as possible. Since I couldn't find anything in the docs, a little trial and error yielded that these field permissions must be granted, anything less would result in 403 when uploading files:

<img width="607" alt="image" src="https://user-images.githubusercontent.com/1450947/183652009-cbcb3b7d-46cb-409a-bf2e-f64d48339901.png">

Sure, `title` is necessary for the file library to stay sane.

However, it seems (at least to me) that `storage` and `mime_type` don't have to be required. (Well, I'm not setting them when calling the api anyway. And I feel a bit uneasy to grant permissions on fields that did not even cause problems when absent in the api request.)

And `filename_download` can't even be set right now. https://github.com/directus/directus/issues/14975

**Comment by azrikahar** [comment]
Opted to move my answer here since it makes sense to separate this out from #14975 into a discussion!

---

> I wonder whether the documentation on uploading files could be improved slightly? (Because I can't know for sure whether there are bugs involved.)
>
> Specifically, it would be nice to have more clarify on the minimal field permissions that have to be set (regarding uploading files).

@andy1li That's a very good question! The need for minimal field permissions for files definitely makes sense. Apart from documenting it, perhaps an alternative would be to do something similar that's already done to `directus_users`, where some fields _are_ the minimal fields by default.

This way it's more obvious that certain fields are integral for files to function properly:

![chrome_bTTAdu8n8D](https://user-images.githubusercontent.com/42867097/183655094-272bbc6b-0064-496c-ba4e-65ca184619b2.png)

**Comment by andy1li** [comment]
https://github.com/directus/directus/issues/14982#issuecomment-1209421170

> Granted I understand how they can be perceived as unnecessary, but I believe we can't exactly deem them as such without further research (maybe aside from Mime Type for certain use cases).

> Let's continue this discussion in #14979 just so we can keep everything in one place.

Thank you, @azrikahar, for the clarification!

I was naive to think that since not setting them in API calls does not cause any problem (apparently they have some default values), it must mean that permissions on them are not necessary. My bad.

**Comment by rijkvanzanten** [comment]
The API will pre-fill some fields based on the files mime types. I think those fields are currently set using the permissions of the user that uploaded the file. If that's the case, we could update the file service used for that metadata update to a full-access service instead, removing the need for some of those specific field permissions, which in turn cleans up some of this confusion 🙂

**Comment by andy1li** [comment]
https://github.com/directus/directus/issues/14975#issuecomment-1209449462

> That's currently by design. You can't set the filename_disk, as Directus adjusts the filename before saving it to disk, so setting that manually will cause various conflicts. Filename download is intended to match the original filename on upload. Happy to discuss improvements on this setup on a Discussion, but closing this for now as it's working as intended

Thank you for the explanation, @rijkvanzanten!

I get it that filename_disk is working as intended. (Didn't really need to change `filename_disk` anyway, just experimenting since there're no docs on fields permissions for file uploading, thus this main thread.)

It seems reasonable that `filename_download` is set as `'blob'`, because I did upload a blob. 

I wonder if there's something I can do to the blob before uploading, so that directus can fill something (more meaningful and more controllable) in `filename_download`? 

EDIT:

To answer my own question, this works:

```javascript
formData.append('file', blob, title)
```

**Reply by rijkvanzanten** [reply]
> It seems reasonable that filename_download is set as 'blob', because I did upload a blob.

This should match filename of the uploaded asset. 

> To answer my own question, this works:
> formData.append('file', blob, title)

Yeah! That sets the "filename" of the blob you're uploading 🙂
